File: engine/unreal/AlienBase_AirlockController.cpp
#include "AlienBase_AirlockController.h"
#include "Engine/World.h"
#include "GameFramework/Actor.h"
#include "Kismet/GameplayStatics.h"

#include "AlienBase_Volume_HubFloorGas.h"
#include "AlienBase_Volume_SublevelAcid.h"

AAlienBase_AirlockController::AAlienBase_AirlockController()
{
    PrimaryActorTick.bCanEverTick = true;

    ArmingDurationSec = 5.0f;
    ActiveDurationSec = 12.0f;
    CooldownDurationSec = 30.0f;

    EventInstigatorTeam = INDEX_NONE;
    TimeInState = 0.0f;
    CurrentState = EAirlockState::Idle;
}

void AAlienBase_AirlockController::BeginPlay()
{
    Super::BeginPlay();
    EnterState(EAirlockState::Idle);
}

void AAlienBase_AirlockController::Tick(float DeltaSeconds)
{
    Super::Tick(DeltaSeconds);
    UpdateState(DeltaSeconds);
}

void AAlienBase_AirlockController::RequestTriggerActivation(FName TriggerId, int32 InstigatorTeam)
{
    if (CurrentState != EAirlockState::Idle)
    {
        return;
    }

    EventInstigatorTeam = InstigatorTeam;
    EnterState(EAirlockState::Arming);

    // TODO: Play global arming VO/sfx and start warning lights.
}

void AAlienBase_AirlockController::EnterState(EAirlockState NewState)
{
    CurrentState = NewState;
    TimeInState = 0.0f;

    switch (CurrentState)
    {
    case EAirlockState::Idle:
        OnEnteredIdle();
        break;
    case EAirlockState::Arming:
        OnEnteredArming();
        break;
    case EAirlockState::Active:
        OnEnteredActive();
        break;
    case EAirlockState::Cooldown:
        OnEnteredCooldown();
        break;
    default:
        break;
    }
}

void AAlienBase_AirlockController::UpdateState(float DeltaSeconds)
{
    TimeInState += DeltaSeconds;

    switch (CurrentState)
    {
    case EAirlockState::Idle:
        break;

    case EAirlockState::Arming:
        if (TimeInState >= ArmingDurationSec)
        {
            EnterState(EAirlockState::Active);
        }
        break;

    case EAirlockState::Active:
        if (TimeInState >= ActiveDurationSec)
        {
            EnterState(EAirlockState::Cooldown);
        }
        break;

    case EAirlockState::Cooldown:
        if (TimeInState >= CooldownDurationSec)
        {
            EnterState(EAirlockState::Idle);
        }
        break;

    default:
        break;
    }
}

void AAlienBase_AirlockController::OnEnteredIdle()
{
    SetHazardVolumesActive(false);
    EventInstigatorTeam = INDEX_NONE;
}

void AAlienBase_AirlockController::OnEnteredArming()
{
    SetHazardVolumesActive(false);
    // TODO: Start sirens, flicker lights, pre-gas vent FX.
}

void AAlienBase_AirlockController::OnEnteredActive()
{
    SetHazardVolumesActive(true);
    // TODO: Update VO/UI to "Airlock sealed / gas released".
}

void AAlienBase_AirlockController::OnEnteredCooldown()
{
    SetHazardVolumesActive(false);
    // TODO: Venting VO/sfx, restore normal ambience.
}

void AAlienBase_AirlockController::SetHazardVolumesActive(bool bActive)
{
    if (HubFloorGasVolume)
    {
        HubFloorGasVolume->SetVolumeActive(bActive);
    }

    if (SublevelAcidVolume)
    {
        SublevelAcidVolume->SetVolumeActive(bActive);
    }
}

File: engine/unity/AlienBaseAirlockController.cs
using UnityEngine;

/// <summary>
/// AlienBaseAirlockController
/// 
/// Map-level Airlock/Gas state machine for 04_Multiplayer_Alien_Base.
/// Controls hazard volumes and responds to trigger consoles.
/// </summary>
public class AlienBaseAirlockController : MonoBehaviour
{
    public enum AirlockState
    {
        Idle,
        Arming,
        Active,
        Cooldown
        }

    [Header("Hazard Volumes")]
    public HazardVolume hubFloorGasVolume;
    public HazardVolume sublevelAcidVolume;

    [Header("Timing (seconds)")]
    public float armingDurationSec = 5.0f;
    public float activeDurationSec = 12.0f;
    public float cooldownDurationSec = 30.0f;

    [Header("Debug")]
    public AirlockState currentState = AirlockState.Idle;
    public float timeInState = 0.0f;
    public int eventInstigatorTeam = -1;

    void Start()
    {
        EnterState(AirlockState.Idle);
    }

    void Update()
    {
        UpdateState(Time.deltaTime);
    }

    public void RequestTriggerActivation(string triggerId, int instigatorTeam)
    {
        if (currentState != AirlockState.Idle)
        {
            return;
        }

        eventInstigatorTeam = instigatorTeam;
        EnterState(AirlockState.Arming);

        // TODO: Play global arming VO/sfx and start warning lights.
    }

    void EnterState(AirlockState newState)
    {
        currentState = newState;
        timeInState = 0.0f;

        switch (currentState)
        {
            case AirlockState.Idle:
                OnEnteredIdle();
                break;
            case AirlockState.Arming:
                OnEnteredArming();
                break;
            case AirlockState.Active:
                OnEnteredActive();
                break;
            case AirlockState.Cooldown:
                OnEnteredCooldown();
                break;
        }
    }

    void UpdateState(float deltaTime)
    {
        timeInState += deltaTime;

        switch (currentState)
        {
            case AirlockState.Idle:
                break;

            case AirlockState.Arming:
                if (timeInState >= armingDurationSec)
                {
                    EnterState(AirlockState.Active);
                }
                break;

            case AirlockState.Active:
                if (timeInState >= activeDurationSec)
                {
                    EnterState(AirlockState.Cooldown);
                }
                break;

            case AirlockState.Cooldown:
                if (timeInState >= cooldownDurationSec)
                {
                    EnterState(AirlockState.Idle);
                }
                break;
        }
    }

    void OnEnteredIdle()
    {
        SetHazardVolumesActive(false);
        eventInstigatorTeam = -1;
    }

    void OnEnteredArming()
    {
        SetHazardVolumesActive(false);
        // TODO: Sirens, lights, pre-gas FX.
    }

    void OnEnteredActive()
    {
        SetHazardVolumesActive(true);
        // TODO: Update VO/UI to "Airlock sealed / gas released".
    }

    void OnEnteredCooldown()
    {
        SetHazardVolumesActive(false);
        // TODO: Venting FX and ambience restore.
    }

    void SetHazardVolumesActive(bool active)
    {
        if (hubFloorGasVolume != null)
        {
            hubFloorGasVolume.SetActive(active);
        }

        if (sublevelAcidVolume != null)
        {
            sublevelAcidVolume.SetActive(active);
        }
    }
}
